### ISSUE TITLE
Storage targets dynamic prefix

### ISSUE THREAD

**Issue by boxedcode** [kubedb/project#243]
With the snapshot feature, is it possible to have a dynamic prefix for the storage targets that support a prefix. For example, with the S3 storage target, you can define the spec like this:

```
s3:
    endpoint: 's3.amazonaws.com'
    bucket: kubedb-qa
    prefix: demo
```

So instead of having a fixed prefix like ```demo``` above, the prefix will be based on the date that the snapshot is being taken, maybe using a customisable prefix template or fixed prefix template options e.g.:

```
s3:
    endpoint: 's3.amazonaws.com'
    bucket: kubedb-qa
    prefixType: date
    prefix: YYYY/MM/dd
```

OR

```
s3:
    endpoint: 's3.amazonaws.com'
    bucket: kubedb-qa
    prefix: template_date
```

What are your thoughts on the above?

**Comment by tamalsaha** [comment]
Currently the uploaded snapshot file will be stored on a path like below:
```
s3://{prefix}/kubedb/{namespace}/{database-name}/{snapshot-name}
```

If you use the scheduled backup approach, it the `snapshot-name` will have timestamp suffixed to it. Does that achieve your desired result?

We are also planning to move to [restic](https://github.com/kubedb/project/issues/168) as uploader. So, this will mean there will be a single archive and no timestamp-ed files.

**Comment by boxedcode** [comment]
> If you use the scheduled backup approach, it the snapshot-name will have timestamp suffixed to it. Does that achieve your desired result?

Not quite, as the ability to upload to a timestamped object key will allow some degree of organisation of the snapshot files.

In addition, historically AWS did not have a good way of navigating S3 buckets that had 100's of objects contained within, whether manually through the console or programmatically. That has since been improved, but I am an advocate of fewer files in an object key path when it comes to S3.

**Comment by tamalsaha** [comment]
Insightful observation @boxedcode ! What do you think about using something like restic to store the backups?

**Comment by boxedcode** [comment]
I've had a brief look at [restic](https://restic.net/) and from what I've seen, it certainly ticks the right boxes from an efficiency and security perspective, and appears to be quite a powerful tool. It is certainly a good next step for kubedb.

**Comment by tamalsaha** [comment]
This will be addressed by #168 .
